### PR TITLE
[RFR] Feature draw from file

### PR DESCRIPTION
**Pull request by adguernier**
[See Trello ](https://trello.com/c/iD8RP3F2/1-1-as-hayao-i-want-to-see-a-particular-game-of-kamon)

[feature_cli_kamon.webm](https://user-images.githubusercontent.com/7949510/232544173-4399f440-2986-40d9-9dea-2781fba1d520.webm)

- [x] init the project with a POC to draw the board in the terminal with the data (2-dimensional array and "undefined" positions)
- [x] refactor the code to organize it and use TypeScript fonctionnality
- [x] fix the feedback

### CODE REVIEW

File: src/board/board.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *  @description Main entrypoint of kamon.js script.
+ *  It will store states of current game, display it, check victory conditions, etc.
+ */
+
+/**
+ * @description This will draw a line of tiles in console from an array of tiles.
+ */
+const drawLine = (lineArray: Array<tileDataInterface>) => {
+    let line = ''        
+    lineArray.forEach(dataTile => {
```
Comment (jonathan-marmelab) [inline src/board/board.ts:18]: Why using `forEach` instead of `map` here ?

File: src/tile/tile.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    A = 'A',
+    B = 'B', 
+    C = 'C', 
+    D = 'D', 
+    E = 'E',
+    F = 'F'
+}
+
+export interface tileDataInterface {
+    color: number,
```
Comment (jonathan-marmelab) [inline src/tile/tile.ts:22]: Why not use `Color` type here ?

File: src/tile/tile.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    symbol: string
```
Comment (jonathan-marmelab) [inline src/tile/tile.ts:23]: Why not use `Symbol` type here ?

File: src/tile/tile.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import chalk from "chalk";
+
+enum Colors {
+    'yellow',
```
Comment (jonathan-marmelab) [inline src/tile/tile.ts:4]: You do not need the single quotes here :)

You should also add values to you enums e.g.

```typescript
enum Colors {
    Yellow = 0,
    // ...
}
```

File: src/index.ts [hunk 5]
```diff
@@ -1 +1,48 @@
...
+import chalk from "chalk";
+import { drawBoard } from './board/board';
+
+let data = [
+    [undefined, {color: 1, symbol: 'C'}, {color: 2, symbol: 'A'}],
+    [{color: 3, symbol: 'A'}, {color: 3, symbol: 'C'}, {color: 2, symbol: 'B'}],
+    [{color: 2, symbol: 'C'}, {color: 1, symbol: 'A'}, undefined]
+];
+
+/*let isFileOk = true;
```
[the remark this answers is not in the document]
Comment (adguernier) [reply src/index.ts:11]: Because it does not work but we wanted to keep it for the next feature

File: package.json [hunk 6]
```diff
@@ -10,7 +10,7 @@
     "@types/blessed": "^0.1.17",
     "babel-jest": "^27.0.6",
     "blessed": "^0.1.81",
-    "chalk": "^5.2.0",
+    "chalk": "^4.1.2",
```
Comment (Nitix) [inline package.json:13]: Open question: 
Can't you upgrade typescript instead? It's better to start without technical debt

File: src/tile/tile.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+export interface tileDataInterface {
+    color: number,
+    symbol: string
+}
+
+/**
+ * @description Draw text from a tile object.
+ */
+export const drawChalkText = (tileData: tileDataInterface): string => {
+    const drawTextFunction = chalk[Colors[tileData.color]];
```
Comment (jonathan-marmelab) [inline src/tile/tile.ts:30]: If you use an enum for `Color` and set the `tileData.color` type to `Color`, this code can be simplfied to `chalk[tileData.color]` which is more readable imho

File: src/board/board.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import chalk from "chalk";
+import {drawChalkText, tileDataInterface} from "../tile/tile";
```
Comment (fzaninotto) [inline src/board/board.ts:2]: interface names should start with a capital letter. Besides, we don't need to put the type in the name, so it should be `TileData`.

File: src/board/board.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+type boardData = Array<Array<undefined|tileDataInterface>>
```
Comment (fzaninotto) [inline src/board/board.ts:4]:
```suggestion
type BoardData = Array<Array<TileData | undefined>>
```

good practice: put `undefined` last

File: src/board/board.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+const BLANK_CHAR = chalk.black(' ')
+
+/**
+ *  @description Main entrypoint of kamon.js script.
+ *  It will store states of current game, display it, check victory conditions, etc.
+ */
+
+/**
+ * @description This will draw a line of tiles in console from an array of tiles.
+ */
```
Comment (fzaninotto) [inline src/board/board.ts:15]: useless comment. If the function name is clear enough, you don't need it.

File: src/board/board.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+const drawLine = (lineArray: Array<tileDataInterface>) => {
```
Comment (fzaninotto) [inline src/board/board.ts:16]: rename `lineArray` to `lines`. Again, don't put the technical type in the name (the IDE will tell us). Names relate to the model, not to the language.

File: src/board/board.ts (hunk 1 above)
Comment (fzaninotto) [inline src/board/board.ts:18]: what's a `dataTile`? can't you find a better name?

File: src/board/board.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    })
+    
+    return line
+}
+
+/**
+ * @description A function to draw a board from a config json.
+ * @borrows drawLine
+ */
+export const drawBoard = (data: boardData) => {
```
Comment (fzaninotto) [inline src/board/board.ts:33]: `data` is probably the worst possible variable name. You can find a better one. 

It seems you're actually receiving an array of lines.

File: src/board/board.ts [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    data.forEach(lineData => {
+        console.log(drawLine(lineData))
+    })
+}
```
Comment (fzaninotto) [inline src/board/board.ts:37]: you have two methods named "drawXXX" but they have very different functions: one builds a string, the other has a side effect (logs to the console). This should be more obvious in their name.

File: src/index.ts [hunk 14]
```diff
@@ -1 +1,48 @@
-console.log("hello, world!");
+const fs = require('fs');
+import chalk from "chalk";
+import { drawBoard } from './board/board';
+
+let data = [
```
Comment (fzaninotto) [inline src/index.ts:5]: If the data doesn't change, replace `let` with `const`

File: src/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (fzaninotto) [reply src/index.ts:11]: We don't commit commented code. Git has the history of past commits.

File: src/tile/tile.ts (hunk 4 above)
Comment (fzaninotto) [reply src/tile/tile.ts:4]: Enums are discouraged in typescript. https://www.executeprogram.com/blog/typescript-features-to-avoid

File: src/tile/tile.ts [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+export interface tileDataInterface {
+    color: number,
+    symbol: string
+}
+
+/**
+ * @description Draw text from a tile object.
+ */
```
Comment (fzaninotto) [inline src/tile/tile.ts:28]: useless description

**Review by fzaninotto — CHANGES_REQUESTED** [review]
this PR lacks unit tests. 

Also, if your PR is finished, please label/prefix it with "RFR" for "Ready for Review". Otherwise, please prefix it with "WIP" for "Work in Progress", and add [a checklist](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/about-task-lists) in the description to show what's done and what's left to do.

File: src/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (adguernier) [reply src/index.ts]: Done

File: src/board/board.ts (hunk 1 above)
Comment (adguernier) [reply src/board/board.ts]: Done

File: src/board/board.ts (hunk 10 above)
Comment (adguernier) [reply src/board/board.ts]: Done

File: src/tile/tile.ts (hunk 15 above)
Comment (adguernier) [reply src/tile/tile.ts]: Done

File: src/board/board.ts (hunk 12 above)
Comment (adguernier) [reply src/board/board.ts:19]: Done

File: src/tile/tile.ts (hunk 4 above)
Comment (adguernier) [reply src/tile/tile.ts:4]: Done

File: src/tile/tile.ts (hunk 3 above)
Comment (adguernier) [reply src/tile/tile.ts:23]: Done

File: src/tile/tile.ts (hunk 7 above)
Comment (adguernier) [reply src/tile/tile.ts]: Done

File: src/board/board.ts (hunk 8 above)
Comment (adguernier) [reply src/board/board.ts]: Done

File: src/board/board.ts (hunk 1 above)
Comment (adguernier) [reply src/board/board.ts]: Done

File: src/board/board.ts (hunk 13 above)
Comment (adguernier) [reply src/board/board.ts:23]: Done

File: src/index.ts (hunk 14 above)
Comment (adguernier) [reply src/index.ts]: Done

File: src/tile/tile.ts (hunk 2 above)
Comment (adguernier) [reply src/tile/tile.ts:22]: Done

File: package.json (hunk 6 above)
Comment (slax57) [reply package.json:13]: Upgrading to chalk v5 seems a bit cumbersome, with very little added value. I recommend keeping v4 for now. However you could upgrade to the latest TS version indeed.

File: src/board/board.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
+import chalk from "chalk";
+import {generateCliText, Tile} from "../tile/tile";
+
+type NullableTile = Tile|undefined
+export type Board = NullableTile[][]
+
+const BLANK_CHAR = chalk.black(' ')
+
+const generateLine = (lines: Tile[]): string => {
```
Comment (slax57) [inline src/board/board.ts:9]:
```suggestion
const renderLine = (lines: Tile[]): string => {
```

File: src/tile/tile.ts [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+
+type Colors = 'yellow' | 'blue' | 'red' | 'green' | 'cyan' | 'white'
+type Symbols = 'A' | 'B' | 'C' | 'D' | 'E' | 'F'
+
+export interface Tile {
+    color: Colors,
+    symbol: Symbols
+}
+
+export const generateCliText = (tile: Tile): string => {
```
Comment (slax57) [inline src/tile/tile.ts:11]:
```suggestion
export const renderTile = (tile: Tile): string => {
```

File: src/tile/tile.ts [hunk 18]
```diff
@@ -0,0 +1,14 @@
+import chalk from "chalk";
+
+type Colors = 'yellow' | 'blue' | 'red' | 'green' | 'cyan' | 'white'
+type Symbols = 'A' | 'B' | 'C' | 'D' | 'E' | 'F'
+
+export interface Tile {
+    color: Colors,
+    symbol: Symbols
+}
```
Comment (slax57) [inline src/tile/tile.ts:9]:
```suggestion
type Color = 'yellow' | 'blue' | 'red' | 'green' | 'cyan' | 'white'
type Symbol = 'A' | 'B' | 'C' | 'D' | 'E' | 'F'

export interface Tile {
    color: Color,
    symbol: Symbol
}
```

File: src/index.ts [hunk 19]
```diff
@@ -1 +1,10 @@
-console.log("hello, world!");
+const fs = require('fs');
```
Comment (slax57) [inline src/index.ts:1]: unused import

File: src/board/board.ts [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+    return lines.reduce((accumulator, tile) => {
+        if (tile == undefined) {
+            return accumulator + BLANK_CHAR
+        }
+        
+        return accumulator + generateCliText(tile)
+    }, line)
+}
+
+export const paintBoard = (data: Board) => {
```
Comment (slax57) [inline src/board/board.ts:20]:
```suggestion
export const renderBoard = (data: Board) => {
```

**Review by slax57 — CHANGES_REQUESTED** [review]
Almost there :slightly_smiling_face:  :muscle:

File: src/tile/tile.ts (hunk 18 above)
Comment (adguernier) [reply src/tile/tile.ts:9]: done

File: src/tile/tile.ts (hunk 17 above)
Comment (adguernier) [reply src/tile/tile.ts:11]: Done

File: src/board/board.ts (hunk 16 above)
Comment (adguernier) [reply src/board/board.ts:9]: Done

File: src/board/board.ts (hunk 20 above)
Comment (adguernier) [reply src/board/board.ts:20]: Done

File: src/index.ts (hunk 19 above)
Comment (adguernier) [reply src/index.ts:1]: done

File: package.json (hunk 6 above)
Comment (adguernier) [reply package.json:13]: Done

**Review by slax57 — APPROVED** [review]

**Pull request merged**